### ISSUE TITLE
`end` is unsafe with user-defined getindex

### ISSUE THREAD

**Issue by mlubin** [JuliaLang/julia#16104]
Ref https://github.com/JuliaOpt/JuMP.jl/issues/730, it seems like there's no way to change the behavior of ``end`` translating to ``size`` when used within a multidimensional array, which means that it's unsafe to use ``end`` to index into JuMP arrays which don't use 1-based indexing. Any hope of a solution to this?

CC @joehuchette @IainNZ 
(Thanks @yuyichao for the pointer)

**Comment by stevengj** [comment]
Shouldn't it call `endof` instead?

**Comment by mlubin** [comment]
@stevengj that's the case only for one-dimensional arrays IIRC

**Comment by stevengj** [comment]
It would be natural to extend it to call `endof(A, i)` in the `i`-th dimension.

**Comment by mlubin** [comment]
I can play around with a PR for that if it has a chance of going through

**Comment by stevengj** [comment]
Here is the [relevant code in julia-syntax.scm](https://github.com/JuliaLang/julia/blob/e5139a7f5f06bf33e507ea4607e198c0fe975965/src/julia-syntax.scm#L93-L110).

Replacing `size(A,i)` with `endof(A,i)` looks pretty easy, and makes sense to me.

A complication is that it calls [trailingsize](base/abstractarray.jl) for the last index in cases with 2 or more indices, so that you can do e.g. `A[i,end]` even if `A` is 3-dimensional.  (Why do we allow this?)   But I guess you could overload `Base.trailingsize` in JuMP too.

**Comment by mlubin** [comment]
``trailingsize`` isn't an issue for us because we already error on ``getindex`` when the number of dimensions doesn't match.

**Comment by StefanKarpinski** [comment]
> A complication is that it calls [trailingsize](https://github.com/JuliaLang/julia/blob/e5139a7f5f06bf33e507ea4607e198c0fe975965/base/abstractarray.jl#L64-L71) for the last index in cases with 2 or more indices, so that you can do e.g. `A[i,end]` even if `A` is 3-dimensional and get `A[i,end,end]`.  (@StefanKarpinski, why do we allow this?)   But I guess you could overload `Base.trailingsize` in JuMP too.

Isnt' this "generalized linear indexing", on the chopping block in 0.5, see https://github.com/JuliaLang/julia/issues/14770.

**Comment by stevengj** [comment]
@mlubin, you would still have to overload it because otherwise it will return `size(A,i)` for the last dimension.  Alternatively, we can call `trailingendof` instead of `trailingsize`, defined as:
```jl
trailingendof(A,n) = n == ndims(A) ? endof(A,n) : trailingsize(A,n)
```
Good to hear that this is on the chopping block, though; we can then remove `trailingendof` and just call `endof(A,n)`.

(So far, the only registered packages that use the unexported `Base.trailingsize` function seem to be DSP.jl, ZipFile.jl, and Zlib.jl.)

**Comment by mlubin** [comment]
Will it be easier to fix this after https://github.com/JuliaLang/julia/issues/14770?

**Comment by stevengj** [comment]
@mlubin, no, I think the two can be done independently if you define `trailingendof` as I suggested above.   After general linear indexing is removed, then `trailingendof` can be removed and replaced with `endof`.

**Comment by mbauman** [comment]
How about we just lower to one function name in all cases?  This moves the logic into a place where custom types can intercept it… and it decouples the last index from the size.  You just need a three-argument function `lastindex(A, dim, is_last_index)`.  Maybe that could even get folded into `endof`.

Cartesian indices still pose a problem: `rand(3,2,1)[CartesianIndex((3,2)), end]`… but that's a much harder problem that may be just fine left unresolved.

**Comment by stevengj** [comment]
@mbauman, because the goal should be to get rid of `trailingendof`.   Anyone overriding `endof` should only have to implement the 2-argument version.
